Checkin for abc remap flow. Files moved from private repo

mathursanjiv opened the pull request:
New command "restructure" to call ABC for remap
-Internally calls resizer to create blob
-Writes blif
-runs ABC
-reads back blif
-Connects constants to tie cells

Updated scripts directory to have another PR

Additional contributor: @ahmadelrouby

maliberty commented on src/rmp/include/rmp/blifParser.h:
```diff
@@ -0,0 +1,60 @@
+ public:
+  BlifParser() { combCount = 0; flopCount = 0; currentInstanceType=""; currentGate = ""; }
+  bool parse(std::string s);
+  void addInput(std::string s) { inputs.push_back(s); }
+  void addOutput(std::string s) { outputs.push_back(s); }
+  void addClock(std::string s) { clocks.push_back(s); }
```
take strings by const & not value to avoid copying; here and throughout

maliberty commented on src/rmp/include/rmp/blifParser.h:
```diff
@@ -0,0 +1,60 @@
+  void addInput(std::string s) { inputs.push_back(s); }
+  void addOutput(std::string s) { outputs.push_back(s); }
+  void addClock(std::string s) { clocks.push_back(s); }
+  void addNewInstanceType(std::string s){
```
give a more meaningful name than 's' (here & throughout)

maliberty commented on src/rmp/messages.txt:
```diff
@@ -0,0 +1,6 @@
+RMP 0001 blif.cpp:73               cannot open file %s
+RMP 0001 blif.cpp:271              cannot open file %s
+RMP 0001 blif.cpp:293              cannot open file %s
+RMP 0001 blif.cpp:309              blif parsed successfully, destroying %d existing instances...
+RMP 0001 blif.cpp:316              inserting %d new instances...
```
You need to use {} in your messages, see the logger doc

maliberty commented on src/rmp/src/blif.cpp.my:
```diff
@@ -0,0 +1,389 @@
+/////////////////////////////////////////////////////////////////////////////
```
What is this file for?

mathursanjiv replied:
Came by mistake. Removed.

maliberty commented on src/rmp/src/rmp.tcl:
```diff
@@ -0,0 +1,101 @@
+##
+############################################################################
+
+# Restructureing could be done in various modes targeting area or timing.
```
typo Restructureing

maliberty commented on src/rmp/src/rmp.tcl:
```diff
@@ -0,0 +1,101 @@
+# Possible values of mode_name: area, delay
+# depth and slack thresholds can be provided for delay mode
+
+sta::define_cmd_args "restructure" { \
+                                      [-slack_threshold slack]\
+                                      [-depth_threshold depth]\
+                                      [-liberty_file liberty_file]\
+                                      [-mode mode_name]\
+                                      [-locell tielow_cell]\
+                                      [-loport tielow_port]\
+                                      [-hiport tiehigh_port]\
+                                      [-hicell tiehigh_cell]
+                                    }
```
Where is the documentation for this module?  I don't know what these arguments are for and neither will any user.

mathursanjiv replied:
added rmp/README.md

mathursanjiv replied:
Tried to re-use argument names used by yosys for tie cell specification. Thought would be easier for user if we use similar names.

maliberty replied:
That's a good thought but I'm more interested in clarity & usability than matching yosys.  repair_tie_fanout take a lib/cell/pin argument in the style of get_lib_pins

mathursanjiv replied:
I'll add comments in tcl file to understand usability. repair_tie_fanout needs additional packaging of string and unpackaging in command, I found separate cell and pin name more usable.

jjcherry56 replied:
I find the command arg names to be both opaque and verbose. The packaging/unpackaging makes the command easier to use. Your choice makes the coding easier, which is not the goal.

mathursanjiv replied:
Updated command interface

maliberty reviewed (changes requested):
I'm going to suspend reviewing until you've documented the functionality and cleaned up the coding style issues.

QuantamHD commented:
How will this command handle multi output cells like adders?

mathursanjiv commented:
> How will this command handle multi output cells like adders?

Multi output combinational cells would be passed to ABC as well, as part of logic cloud.

arlpetergadfort commented:
> > How will this command handle multi output cells like adders?
> 
> Multi output combinational cells would be passed to ABC as well, as part of logic cloud.

Have you tested with adders in the design? ABC cannot handle multi-output cells.

ahmadelrouby commented on src/OpenDB/test/cpp/helper.cpp:
```diff
@@ -94,3 +94,41 @@ dbDatabase* create2LevetDbWithBTerms()
 }
+
+
+dbDatabase* create3LevetDbNoBTerms()
```
added this when blif resided in opendb to test the writer but now it's no longer needed after moving the code so I'll just remove it.

mathursanjiv commented:
> > > How will this command handle multi output cells like adders?
> > 
> > 
> > Multi output combinational cells would be passed to ABC as well, as part of logic cloud.
> 
> Have you tested with adders in the design? ABC cannot handle multi-output cells.

Yes, we pass, for example, HA to ABC in blif, it does not complain. The netlist coming out from blif does not have HA. It decomposes it to single output gates. Formal equivalence between ABC in and out is clean.
Do you think this is not expected?

jjcherry56 commented on src/rmp/README.md:
```diff
@@ -0,0 +1,39 @@
+
+Argument Description
+- ``liberty_file`` Liberty file with description of cells used in design. This would be passed to ABC.
+- ``mode`` could be area or delay. In area mode focus is area reduction and timing may degrade. In delay mode delay would be reduced but area may increase.
```
-mode area|delay is the correct way to indicate the command has 2 possible values. not "could be".

mathursanjiv replied:
Done

jjcherry56 commented on src/rmp/README.md:
```diff
@@ -0,0 +1,39 @@
+- ``mode`` could be area or delay. In area mode focus is area reduction and timing may degrade. In delay mode delay would be reduced but area may increase.
+- ``-slack_threshold`` specifies slack value below which timing paths need to be analyzed for restructuring
+- ``-depth_threshold`` specifies the path depth above which a timing path would be considered for restructuring
+- ``locell`` specifies tie cell which can drive constant zero
```
look at the repair_tie_fanout and use  consistent argument names

mathursanjiv replied:
Used interface of repair_tie_fanout to take lib/cell/pin format

jjcherry56 commented on src/rmp/include/rmp/Restructure.h:
```diff
@@ -0,0 +1,126 @@
+using utl::Logger;
+
+typedef enum {
+  AREA_1_MODE = 0,
+  AREA_2_MODE,
+  AREA_3_MODE,
+  DELAY_1_MODE,
+  DELAY_2_MODE,
+  DELAY_3_MODE,
+  DELAY_4_MODE
```
_MODE is only redundant if you use enum class instead of the ancient C version with typedef because the values are global. use enum class instead.

mathursanjiv replied:
Removed

jjcherry56 commented on src/rmp/include/rmp/Restructure.h:
```diff
@@ -0,0 +1,126 @@
+  ord::OpenRoad* openroad_;
+  Logger* logger_;
+  std::string logfile_ = "";
+  std::string locell_ = "";
+  std::string loport_ = "";
+  std::string hicell_ = "";
+  std::string hiport_ = "";
```
initing an std::string to "" is pointless

jjcherry56 commented on src/rmp/include/rmp/Restructure.h:
```diff
@@ -0,0 +1,130 @@
+  std::string hiport_ = "";
+
+  // db vars
+  sta::dbSta* open_sta_ = nullptr;
```
do no init variable in the structure definition. that is what a constructor is for. these variables in particular are inited in the initRestructure function as part of the openroad tool protocol so they do not need to be explicitly inited

jjcherry56 commented on src/rmp/include/rmp/Restructure.h:
```diff
@@ -0,0 +1,130 @@
+
+  std::string input_blif_file_name_;
+  std::string output_blif_file_name_;
+  std::vector<std::string> lib_file_names_;
```
you can get the lib file names from opensta so there is no point in having a local copy of them. api, not structure

mathursanjiv replied:
Getting lib files from OpenSTA now. USing this variable to pass from OpenSTA to multiple dumps of ABC control script

jjcherry56 commented on src/rmp/include/rmp/Restructure.h:
```diff
@@ -0,0 +1,130 @@
+  std::vector<std::string> lib_file_names_;
+  std::set<odb::dbInst*> path_insts_;
+
+  Mode opt_mode_ = DELAY_1_MODE;
```
again, bad to init a value in the structure. the command default value should override and this becomes meaningless

jjcherry56 commented on src/rmp/include/rmp/blifParser.h:
```diff
@@ -0,0 +1,75 @@
+  void addInput(std::string& input) { inputs.push_back(input); }
+  void addOutput(std::string& output) { outputs.push_back(output); }
+  void addClock(std::string& clock) { clocks.push_back(clock); }
+  void addNewInstanceType(std::string& type)
```
but code in .cpp files, not class definitions

jjcherry56 commented on src/rmp/include/rmp/blifParser.h:
```diff
@@ -0,0 +1,75 @@
+  std::vector<std::string> inputs;
+  std::vector<std::string> outputs;
+  std::vector<std::string> clocks;
+  std::vector<std::tuple<std::string, std::string, std::vector<std::string>>>
```
this is a mouthful and deserves a typedef for readability

jjcherry56 commented on src/rmp/src/blifParser.cpp:
```diff
@@ -0,0 +1,120 @@
+#include <string>
+#include <vector>
+
+namespace blifParserClass {
```
blif_parser would be more appropriate. no camel case for namespace, and they should not be so verbose

jjcherry56 commented on src/rmp/src/rmp.tcl:
```diff
@@ -0,0 +1,109 @@
+##
+############################################################################
+
+# Restructuring could be done in various modes targeting area or timing.
```
various modes is vague and ambiguous

jjcherry56 commented on src/rmp/src/rmp.tcl:
```diff
@@ -0,0 +1,109 @@
+                                      [-slack_threshold slack]\
+                                      [-depth_threshold depth]\
+                                      [-liberty_file liberty_file]\
+                                      [-mode mode_name]\
```
mode is a vague term to use.
target? 
instead of mode_name it should list the valid values; area|timing|size or whatever

mathursanjiv replied:
Changed mode to target

jjcherry56 commented on src/rmp/src/rmp.tcl:
```diff
@@ -0,0 +1,109 @@
+# Argument Description
+# liberty_file:    Liberty file with description of cells used in design. This would be passed to ABC.
+# mode:            Could be area or delay. In area mode focus is area reduction and timing may degrade. In delay mode delay would be reduced but area may increase.
+# slack_threshold: specifies slack value below which timing paths need to be analyzed for restructuring
```
awkward.
timing path minimum slack for restructuring
what does 'need to be analyzed' mean? maybe it will  restructure it if it has less than that slack?

mathursanjiv replied:
Yes, all paths violating minimum slack do not qualify unless they exhibit path depth issues as well. A violating timing path with say path depth of 2 would not be re-synthesized. The path depth threshold can be controlled by an argument as well however.

jjcherry56 replied:
The point is not to explain it to me. It is to explain it to the sure in the documentation and this does not do an adequate job.

mathursanjiv replied:
Updated descrition

jjcherry56 commented on src/rmp/src/blifParser.cpp:
```diff
@@ -0,0 +1,120 @@
+void setNewInput(std::string input, rmp::BlifParser* parser)
+{
+  if (input == "\\")
+    return;
```
there is no reason to use return here (ie, non-local exit)

jjcherry56 commented on src/rmp/src/blifParser.cpp:
```diff
@@ -0,0 +1,120 @@
+void setNewOutput(std::string output, rmp::BlifParser* parser)
+{
+  if (output == "\\")
+    return;
```
ditto

jjcherry56 commented on src/rmp/src/rmp.i:
```diff
@@ -0,0 +1,89 @@
+
+%inline %{
+
+void set_mode_cmd(const char* modeName)
```
what is the point of defining all these little functions? logically these are just arguments to the restructure function

mathursanjiv replied:
removed

arlpetergadfort commented:
> > > > How will this command handle multi output cells like adders?
> > > 
> > > 
> > > Multi output combinational cells would be passed to ABC as well, as part of logic cloud.
> > 
> > 
> > Have you tested with adders in the design? ABC cannot handle multi-output cells.
> 
> Yes, we pass, for example, HA to ABC in blif, it does not complain. The netlist coming out from blif does not have HA. It decomposes it to single output gates. Formal equivalence between ABC in and out is clean.
> Do you think this is not expected?

Since ABC doesn't support the HA, I can see why it would be missing and replaced with something ABC can handle. I think you should see what happens on a larger design. My concern would be that the optimization would simply create 3+ gates for each FA/HA and that will not be an improved result for designs with lots of adders. YOSYS either inserts the adders before ABC cuts the logic at the adder boundaries, or just allows ABC to replace with simple gates. You may need a separate method of handling these gates, or paths that contain multi-output cells, since ABC cannot optimize without replacing with simpler cells.

mathursanjiv commented:
> > > > > How will this command handle multi output cells like adders?
> > > > 
> > > > 
> > > > Multi output combinational cells would be passed to ABC as well, as part of logic cloud.
> > > 
> > > 
> > > Have you tested with adders in the design? ABC cannot handle multi-output cells.
> > 
> > 
> > Yes, we pass, for example, HA to ABC in blif, it does not complain. The netlist coming out from blif does not have HA. It decomposes it to single output gates. Formal equivalence between ABC in and out is clean.
> > Do you think this is not expected?
> 
> Since ABC doesn't support the HA, I can see why it would be missing and replaced with something ABC can handle. I think you should see what happens on a larger design. My concern would be that the optimization would simply create 3+ gates for each FA/HA and that will not be an improved result for designs with lots of adders. YOSYS either inserts the adders before ABC cuts the logic at the adder boundaries, or just allows ABC to replace with simple gates. You may need a separate method of handling these gates, or paths that contain multi-output cells, since ABC cannot optimize without replacing with simpler cells.

We have tried over 100k inst resynthesis in ABC. Preserving HA/FA vs dissolving is debatable. HA/FA are not seen in netlists these days. If we skip logic containing these cells, optimization could be limited considerably.

maliberty commented on src/rmp/src/Restructure.cpp:
```diff
@@ -0,0 +1,555 @@
+          auto new_inst = odb::dbInst::create(block_, const_master, inst_name.c_str());
+          if (new_inst) {
+            odb::dbITerm::disconnect(iterm);
+            odb::dbITerm::connect(new_inst->findITerm(const_port.c_str()), net);
```
It would be cleaner to find the dbMTerm once and use dbInst::getITerm for the mterm rather than using names.

maliberty commented on src/rmp/src/Restructure.cpp:
```diff
@@ -0,0 +1,555 @@
+          // clk_gating_setup, clk_gating_hold
+          false,
+          false);
+  bool skip_top_paths = false;
+  std::size_t path_found = path_ends->size();
+  if (path_found > 1000) {
+    skip_top_paths = true;
+  }
```
I see no use of skip_top_paths

mathursanjiv replied:
Yes, removed

maliberty commented on src/rmp/include/rmp/Restructure.h:
```diff
@@ -0,0 +1,130 @@
+
+enum class Mode
+{
+  AREA_1 = 0,
+  AREA_2,
+  AREA_3,
+  DELAY_1,
+  DELAY_2,
+  DELAY_3,
+  DELAY_4
```
Are there more meaningful names than these?  At least some descriptive comments would help.

mathursanjiv replied:
These are just different versions of area/delay scripts. I can add VERSION to enum name. There is not conceptual name that can be given to scripts as there is command change/reshuffles.

maliberty commented on src/rmp/src/Restructure.cpp:
```diff
@@ -0,0 +1,555 @@
+  for (auto& path_end : *path_ends) {
+    if (opt_mode_ >= Mode::DELAY_1) {
+      sta::PathExpanded expanded(path_end->path(), open_sta_);
+      logger_->report("Got path of depth {}", expanded.size() / 2);
```
This looks more like a debug message.  The message doesn't seem useful to an end user.

mathursanjiv replied:
This is to flag each path violating path depth.

maliberty commented on src/rmp/src/Restructure.cpp:
```diff
@@ -0,0 +1,555 @@
+        ends.insert(path_end->vertex(sta_state)->pin());
+        end_count++;
+      }
+      if (end_count > 5)  // limit blob size for timing
```
why 5?

mathursanjiv replied:
Upto 5 size should be ok. We may have 4 as well. This would become clear once we experiment with timing restructuring on few designs.

maliberty commented on src/rmp/src/Restructure.cpp:
```diff
@@ -0,0 +1,555 @@
+
+  // unconstrained end points
+  auto errors
+      = open_sta_->checkTiming(false, false, false, true, true, false, false);
```
what are these args?

mathursanjiv replied:
Added comments.

maliberty commented on src/rmp/src/Restructure.cpp:
```diff
@@ -0,0 +1,555 @@
+
+  // abc optimization
+  bool area_mode = opt_mode_ <= Mode::AREA_3;
+  std::string abc_script_file = "ord_abc_script.tcl";
```
I don't think we should just write into cwd but should allow the user to give a run directory (eg in flows we would use a subdir for the design).

maliberty commented on src/rmp/src/Restructure.cpp:
```diff
@@ -0,0 +1,555 @@
+  int best_inst_count = std::numeric_limits<int>::max();
+  for (int opt_mode = static_cast<int>(area_mode) ? static_cast<int>(Mode::AREA_1) : static_cast<int>(Mode::DELAY_1);
+       opt_mode <= (static_cast<int>(area_mode) ? static_cast<int>(Mode::AREA_3) : static_cast<int>(Mode::DELAY_4));
+       opt_mode += 1) {
```
opt_mode++ ?

maliberty commented on src/rmp/src/Restructure.cpp:
```diff
@@ -0,0 +1,555 @@
+      std::string abc_command = std::string("yosys-abc < ") + abc_script_file;
+      if (logfile_ != "")
+        abc_command = abc_command + " > " + logfile_;
+      std::system(abc_command.c_str());
```
Check the return code for failure and error as needed.

maliberty commented on src/rmp/src/Restructure.cpp:
```diff
@@ -0,0 +1,555 @@
+      // failed, revert
+
+    } else {
+      // slack improved, accept
```
tbd??

mathursanjiv replied:
Yes, we have not yet worked on timing aspect of re-structuring. I can remove this till we actually get to it.

maliberty commented on src/rmp/src/Restructure.cpp:
```diff
@@ -0,0 +1,555 @@
+    float new_slack;
+    sta::Vertex* worst_vertex;
+    open_sta_->worstSlack(sta::MinMax::max(), new_slack, worst_vertex);
+    if (!area_mode && new_slack < worst_slack) {
```
you already check !area_mode on 214

maliberty commented on src/rmp/src/Restructure.cpp:
```diff
@@ -0,0 +1,555 @@
+    sta::Vertex* worst_vertex;
+    open_sta_->worstSlack(sta::MinMax::max(), new_slack, worst_vertex);
+    if (!area_mode && new_slack < worst_slack) {
+      // failed, revert
```
tbd?

maliberty commented on src/rmp/src/Restructure.cpp:
```diff
@@ -0,0 +1,555 @@
+{
+  int const_nets = 0;
+  for (auto block_net : top_block->getNets()) {
+    if (block_net->getSigType() == odb::dbSigType::GROUND
+        || block_net->getSigType() == odb::dbSigType::POWER)
```
block_net->getSigType()->isSupply()

jjcherry56 commented on src/OpenDB/test/data/basic_read.blif:
```diff
@@ -0,0 +1,8 @@
+.model tmp_circuit
```
why are the tests in opendb?
also 3_place.def is way bigger than you should have in a UNIT TEST

jjcherry56 commented on src/rmp/src/rmp.tcl:
```diff
@@ -0,0 +1,126 @@
+sta::define_cmd_args "restructure" { \
+                                      [-slack_threshold slack]\
+                                      [-depth_threshold depth]\
+                                      [-liberty_file liberty_file]\
```
There is no reason for this command to have a liberty file argument. 
Get the liberty filenames from opensta.

mathursanjiv replied:
Removed this option and using OpenSTA now

jjcherry56 commented on src/rmp/test/gcd_restructure.ok:
```diff
@@ -0,0 +1,28 @@
+Number of paths for restructure 70
+Number of end points for restructure 35
+Got 1290 pins in cloud extraction
+Found 422 Insts for restructuring ...
```
don't use abbreviations like "Insts" in messages

jjcherry56 commented on src/rmp/test/gcd_restructure.ok:
```diff
@@ -0,0 +1,28 @@
+[WARNING RMP-0033] -tiehi_pin not specified
+Number of paths for restructure 70
+Number of end points for restructure 35
+Got 1290 pins in cloud extraction
```
WTF is a cloud extraction?

mathursanjiv replied:
What is WTF?

jjcherry56 commented on src/rmp/test/gcd_restructure.ok:
```diff
@@ -0,0 +1,28 @@
+Optimized to 257 instance in iter 1
+Optimized to 255 instance in iter 2
+[INFO RMP-0005] blif parsed successfully, destroying 422 existing instances...
+[INFO RMP-0006] Found 89 Inputs, 52 Outputs, 0 Clocks, 241 Combinational gates, 0 Flops after parsing the blif file.
```
I would much rather see "register" used everywhere you currently use "flop".

jjcherry56 commented on src/rmp/test/gcd_restructure.tcl:
```diff
@@ -0,0 +1,18 @@
+report_design_area
+
+restructure -liberty_file Nangate45/Nangate45_typ.lib -target area -logfile abc.log
+file delete abc.log
```
why are you writing a log file if you turn around and delete it?

jjcherry56 commented on src/rmp/test/helpers.tcl:
```diff
@@ -0,0 +1,89 @@
+}
+
+set test_dir [file dirname [file normalize [info script]]]
+set result_dir [file join $test_dir "results"]
```
look at all of the regression tests that use the function make_result_file defined in helpers.tcl, which you should link, not copy

jjcherry56 commented on src/rmp/test/helpers.tcl:
```diff
@@ -0,0 +1,89 @@
+# Helper functions common to multiple regressions.
```
This should be a file link to the one in /test. maybe it is, but make sure it is not a copy

jjcherry56 commented on src/rmp/test/test_blif_writer.blif.ok:
```diff
@@ -0,0 +1,8 @@
+
+.gate AND2_X1 A1=inp1 A2=w0 ZN=w1
+.gate OR2_X1 A1=inp1 A2=w1 ZN=w2
+.end
```
see that little circle? it means there is a missing return. that is a bug

ahmadelrouby replied:
I'm not sure I follow. Can you please elaborate?

mathursanjiv replied:
Hi Ahmad, please add newline after last line of blif. Newline is missing after ".end"

jjcherry56 commented on src/utl/include/utl/Logger.h:
```diff
@@ -74,6 +74,7 @@ namespace utl {
     X(STA) \
     X(STT) \
     X(TAP) \
+    X(RMP) \
```
put this is alphabetical order

jjcherry56 commented on src/rmp/test/gcd_restructure.ok:
```diff
@@ -0,0 +1,28 @@
+Got 1290 pins in cloud extraction
+Found 422 Insts for restructuring ...
+[INFO RMP-0002] Blif writer successfully dumped file with 422 instances.
+Optimized to 241 instance in iter 0
```
instances

ahmadelrouby commented on src/rmp/include/rmp/blifParser.h:
```diff
@@ -1,57 +1,45 @@
+  std::vector<std::string>& getInputs();
+  std::vector<std::string>& getOutputs();
+  std::vector<std::string>& getClocks();
+  std::vector<gate_>& getGates();
```
I'm not sure I understand the question, but If you're asking for making these methods const methods which also return const vector references then I did that. But if you meant something else, please let me know.

ahmadelrouby commented on src/rmp/src/blifParser.cpp:
```diff
@@ -0,0 +1,189 @@
+
+void setNewOutput(std::string output, rmp::BlifParser* parser)
+{
+  if (output != "\\")
```
There shouldn't be a case where this actually happens, the boost parser doesn't consume empty string and the parser we defined here specifies that there needs to be at least one output, otherwise, the parser will fail.

ahmadelrouby commented on src/rmp/src/blif.cpp:
```diff
@@ -0,0 +1,533 @@
+      if (connectedIterms.size() == 1) {
+        if (iterm->getIoType() == odb::dbIoType::INPUT)
+          inputs.insert(netName);
+        else if (iterm->getIoType() == odb::dbIoType::OUTPUT)
```
I haven't seen any mention of inouts in the blif spec, so if we hit such an iterm, should we treat as input or output?

jjcherry56 commented on src/rmp/test/gcd_restructure.tcl:
```diff
@@ -10,7 +10,10 @@ report_worst_slack
 report_design_area
 
-restructure -target area -abc_logfile results/abc.log
+set tiehi "NangateOpenCellLibrary/LOGIC1_X1/Z"
```
you don't need the library name

jjcherry56 commented on src/rmp/test/gcd_restructure.tcl:
```diff
@@ -10,7 +10,10 @@ report_worst_slack
+set tiehi "NangateOpenCellLibrary/LOGIC1_X1/Z"
+set tielo "NangateOpenCellLibrary/LOGIC0_X1/Z"
+
+restructure -target area -abc_logfile results/abc_rcon.log  -tielo_pin $tielo -tiehi_pin $tiehi
```
I am not a fan of using pin here. It is what DEF uses (and sdc for that matter) but I consistently use port everywhere outside of required sdc compatibility

jjcherry56 commented on src/rmp/test/gcd_restructure.ok:
```diff
@@ -12,8 +12,7 @@
 Design area 700 u^2 11% utilization.
-[WARNING RMP-0032] -tielo_pin not specified
-[WARNING RMP-0033] -tiehi_pin not specified
+Removed 0 instances with constant outputs...
```
what is the point of the ellipsis at the end of the line?

mathursanjiv commented:
> **parasitices** is not a word

Corrected.

maliberty commented on src/rmp/src/Restructure.cpp:
```diff
@@ -87,31 +86,33 @@ void Restructure::reset()
 {
   reset();
-  block_ = openroad_->getDb()->getChip()->getBlock();
+  block_ = ord::OpenRoad::openRoad()->getDb()->getChip()->getBlock();
```
Avoid using ord::OpenRoad::openRoad().  This can be gotten from db_ instead.

maliberty commented on src/rmp/src/Restructure.cpp:
```diff
@@ -87,31 +86,33 @@ void Restructure::reset()
   sta::Slack worst_slack = slack_threshold;
-  open_sta_ = openroad_->getSta();
+
+  open_sta_ = ord::OpenRoad::openRoad()->getSta();
```
In init()

maliberty commented on src/rmp/src/Restructure.cpp:
```diff
@@ -87,31 +86,33 @@ void Restructure::reset()
 void Restructure::getBlob(unsigned max_depth)
 {
-  rsz::Resizer* resizer = openroad_->getResizer();
+  rsz::Resizer* resizer = ord::OpenRoad::openRoad()->getResizer();
```
In init()

maliberty commented on src/rmp/src/Restructure.cpp:
```diff
@@ -147,7 +148,7 @@ void Restructure::runABC()
              countConsts(block_));
 
-  Blif blif_(openroad_, locell_, loport_, hicell_, hiport_);
+  Blif blif_(ord::OpenRoad::openRoad(), locell_, loport_, hicell_, hiport_);
```
again

maliberty commented on src/rmp/src/Restructure.cpp:
```diff
@@ -202,32 +207,21 @@ void Restructure::runABC()
-      // slack improved, accept
-    }
-  }
+  rsz::Resizer* resizer = ord::OpenRoad::openRoad()->getResizer();
```
again

maliberty commented on src/rmp/src/rmp.tcl:
```diff
@@ -52,19 +51,18 @@
                                       [-target area|timing]\
-                                      [-tielo_pin tielow_pin]\
-                                      [-tiehi_pin tiehigh_pin]
+                                      [-liberty_file liberty_file]\
```
Is this taken from sta now?

jjcherry56 replied:
There should not be a -liberty_file argument to the command

mathursanjiv replied:
Had removed this but had to bring this back since ABC takes only one liberty file. Private regressions are failing since we need more than one liberty file there. So we have to provide "DONT_USE_SC_LIB" just as we do in synth.tcl for ABC.

jjcherry56 replied:
that is disgusting

arlpetergadfort replied:
@mathursanjiv ABC supports multiple liberty files, there is no limitation. The current version of YOSYS the OpenROAD-flow is using is the one that can only handle one liberty file, but that has since been changed. So, there should be no problem providing ABC with more than one.  One issue I do see in using the original liberty files, is the dont_use flags are not set in those and ABC may use cells that should have been marked dont_use.

mathursanjiv replied:
@arlpetergadfort can you point me to some testcase which runs with multiple liberty files? May be its version or command difference. For me ABC could not find master from one of liberty file provided. After using integrated liberty file all regressions went through without that error.

arlpetergadfort replied:
I'm not sure there are any examples at the moment.
Here is the code YOSYS uses to build the abc script: https://github.com/YosysHQ/yosys/blob/862e84eb3d85dfaacc7025b33c7dbe038bfff2d5/passes/techmap/abc.cc#L714
ABC seems to choke on some of the memory liberty files.

Maybe the better approach would be to only select the liberty files that contain non-blocks. Ie. if the liberty file only contains hard macros (or all don't use cells), ABC doesn't need that file. I suspect in most cases you would end up providing just one or two liberty files instead of all of them.

jjcherry56 replied:
I'll say it again: the liberty file should be obvious by looking at the gates that are included in the restructure blob. I do not believe there is any example in all of our tests that have more than one standard cell library.

mathursanjiv replied:
With single lib, gf12/aes failed and few more from private regression.

maliberty commented on src/TritonRoute/messages.txt:
```diff
@@ -0,0 +1,258 @@
+DRT 0000 FlexDR_init.cpp:362       Error: initNetTerms unsupported obj
```
This shouldn't be here - it is removed on master.

jjcherry56 commented on src/rmp/src/Restructure.cpp:
```diff
@@ -62,14 +62,13 @@ namespace rmp {
 void Restructure::makeComponents()
 {
-  db_ = openroad_->getDb();
+  db_ = ord::OpenRoad::openRoad()->getDb();
```
This is essentially referencing a global variable, which is not a good practice.

mathursanjiv commented:
Even with 2 liberty files, it complaints during Blif read for cell not
found even if it's there in liberty.

On Tue, Jun 29, 2021, 19:48 arlpetergadfort ***@***.***>
wrote:

> ***@***.**** commented on this pull request.
> ------------------------------
>
> In src/rmp/src/rmp.tcl
> <https://github.com/The-OpenROAD-Project/OpenROAD/pull/752#discussion_r660664877>
> :
>
> >                                        [-target area|timing]\
> -                                      [-tielo_pin tielow_pin]\
> -                                      [-tiehi_pin tiehigh_pin]
> +                                      [-liberty_file liberty_file]\
>
> I'm not sure there are any examples at the moment.
> Here is the code YOSYS uses to build the abc script:
> https://github.com/YosysHQ/yosys/blob/862e84eb3d85dfaacc7025b33c7dbe038bfff2d5/passes/techmap/abc.cc#L714
> ABC seems to choke on some of the memory liberty files.
>
> Maybe the better approach would be to only select the liberty files that
> contain non-blocks. Ie. if the liberty file only contains hard macros (or
> all don't use cells), ABC doesn't need that file. I suspect in most cases
> you would end up providing just one or two liberty files instead of all of
> them.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/The-OpenROAD-Project/OpenROAD/pull/752#discussion_r660664877>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ANVEL6NNATMCFRMTT26PJJ3TVHI4JANCNFSM463NGUHQ>
> .
>

arlpetergadfort commented:
> Even with 2 liberty files, it complaints during Blif read for cell not found even if it's there in liberty.

I have not attempted to run anything from this PR, but I know from running ABC its supported. However, there might be something else going wrong. Maybe if you have a sample blif and abc script that might shed some light on it.

mathursanjiv commented:
> > Even with 2 liberty files, it complaints during Blif read for cell not found even if it's there in liberty.
> 
> I have not attempted to run anything from this PR, but I know from running ABC its supported. However, there might be something else going wrong. Maybe if you have a sample blif and abc script that might shed some light on it.

I only have public regression which are all single liberty based. Private regression has problems due to multiple libs which cannot be shared. What command/options have you used to read liberty and which version do you use?

mathursanjiv commented:
> It is idiotic that the flow scripts write sdc without changing anything about the sdc, but that is another topic.
> modify one of your unit tests to show the issue and I'll see what makes sense.
> opensta is incremental so it should see the changes and automatically update but I suspect the blif reader is running around the callbacks or something so it doesn't see the changes.
> write_sdc does not depend on the graph or gate delays, so calliing ensureGraph/ensureLevelized etc makes no sense.

@jjcherry56 I have removed ensureGraph/ensureLevelizer etc. Updated testcase OpenROAD/src/rmp/test/gcd_restructure.tcl to include write_sdc command. Not write_sdc is crashing. Now the crash does not seem to go away even if have those APIs. Please use this branch to reproduce the issue https://github.com/mathursanjiv/OpenROAD/tree/abc_remap
Thanks.

jjcherry56 commented:
It does not compile with clang:
In file included from /Users/cherry/openroad/src/rmp/src/blif.cpp:36:
/Users/cherry/openroad/src/rmp/src/../include/rmp/blif.h:82:15: error: implicit
      instantiation of undefined template 'std::__1::basic_string<char>'
  std::string const0_cell_;
              ^
/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX11.3.sdk/usr/include/c++/v1/iosfwd:209:32: note: 
      template is declared here
    class _LIBCPP_TEMPLATE_VIS basic_string;
                               ^
In file included from /Users/cherry/openroad/src/rmp/src/blif.cpp:36:
/Users/cherry/openroad/src/rmp/src/../include/rmp/blif.h:83:15: error: implicit
      instantiation of undefined template 'std::__1::basic_string<char>'
  std::string const0_cell_port_;
              ^
/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX11.3.sdk/usr/include/c++/v1/iosfwd:209:32: note: 
      template is declared here
    class _LIBCPP_TEMPLATE_VIS basic_string;
                               ^
In file included from /Users/cherry/openroad/src/rmp/src/blif.cpp:36:
/Users/cherry/openroad/src/rmp/src/../include/rmp/blif.h:84:15: error: implicit
      instantiation of undefined template 'std::__1::basic_string<char>'
  std::string const1_cell_;
              ^
/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX11.3.sdk/usr/include/c++/v1/iosfwd:209:32: note: 
      template is declared here
    class _LIBCPP_TEMPLATE_VIS basic_string;
                               ^
In file included from /Users/cherry/openroad/src/rmp/src/blif.cpp:36:
/Users/cherry/openroad/src/rmp/src/../include/rmp/blif.h:85:15: error: implicit
      instantiation of undefined template 'std::__1::basic_string<char>'
  std::string const1_cell_port_;
              ^
/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX11.3.sdk/usr/include/c++/v1/iosfwd:209:32: note: 
      template is declared here
    class _LIBCPP_TEMPLATE_VIS basic_string;
                               ^
4 errors generated.

jjcherry56 commented on src/rmp/src/blifParser.cpp:
```diff
@@ -0,0 +1,194 @@
+  parser->endParser();
+}
+
+template <typename Iterator>
```
I think using a template here is overkill. Just declare the args to be the type used in the one call.

jjcherry56 commented on src/rmp/test/blif_reader.blif:
```diff
@@ -0,0 +1,9 @@
+.gate AND2_X1 A1=n1 A2=n2 ZN=n5
+.gate AND2_X1 A1=n3 A2=n4 ZN=n6
+.gate OR2_X1 A1=n5 A2=n6 ZN=n7
+.end
```
missing eol

jjcherry56 commented on src/rmp/test/blif_reader.tcl:
```diff
@@ -0,0 +1,32 @@
+}
+
+puts "pass"
+exit
```
never need exit in test scripts

jjcherry56 commented on src/rmp/test/blif_reader_const.blif:
```diff
@@ -0,0 +1,11 @@
+.gate AND2_X1 A1=n1 A2=lonet ZN=n5
+.gate AND2_X1 A1=n3 A2=hinet ZN=n6
+.gate OR2_X1 A1=n5 A2=n6 ZN=n7
+.end
```
missing eol

jjcherry56 commented on src/rmp/test/blif_reader_const.tcl:
```diff
@@ -0,0 +1,32 @@
+}
+
+puts "pass"
+exit
```
never need exit in test scripts

jjcherry56 commented on src/rmp/test/blif_reader_sequential.blif:
```diff
@@ -0,0 +1,10 @@
+.gate AND2_X1 A1=n3 A2=n4 ZN=n6
+.mlatch DFF_X1 D=n7 Q=n8 QN=dummy_0 clk
+.gate OR2_X1 A1=n5 A2=n6 ZN=n7
+.end
```
tired of repeating myself

jjcherry56 commented on src/rmp/test/blif_reader_sequential.tcl:
```diff
@@ -0,0 +1,32 @@
+}
+
+puts "pass"
+exit
```
never need exit in test scripts

jjcherry56 commented on src/rmp/test/blif_writer.tcl:
```diff
@@ -0,0 +1,26 @@
+}
+
+puts "pass"
+exit
```
never need exit in test scripts

jjcherry56 commented on src/rmp/test/blif_writer_consts.tcl:
```diff
@@ -0,0 +1,26 @@
+}
+
+puts "pass"
+exit
```
never need exit in test scripts

jjcherry56 commented on src/rmp/test/blif_writer_consts.tcl:
```diff
@@ -0,0 +1,26 @@
+
+set isDiff [diff_files "./results/blif_writer_consts.blif" "blif_writer_consts.blif.ok"]
+if {$isDiff != 0} {
+    exit 1
```
should not call exit in tests

jjcherry56 commented on src/rmp/test/blif_writer_hanging.tcl:
```diff
@@ -0,0 +1,26 @@
+
+set isDiff [diff_files "./results/blif_writer_hanging.blif" "blif_writer_hanging.blif.ok"]
+if {$isDiff != 0} {
+    exit 1
```
this is not a shell script with a return code.

jjcherry56 commented on src/rmp/test/blif_writer_input_output.blif.ok:
```diff
@@ -0,0 +1,9 @@
+.gate AND2_X1 A1=n9 A2=n10 ZN=n13
+.gate AND2_X1 A1=n11 A2=n12 ZN=n14
+.gate AND2_X1 A1=n13 A2=n14 ZN=n15
+.end
```
missing eol

jjcherry56 commented on src/rmp/test/const_cell_removal.tcl:
```diff
@@ -0,0 +1,15 @@
+
+report_design_area
+
+exit
```
never need exit in test scripts

jjcherry56 commented:
This message is meaningless to a user:
Size of errors = 1

And this warning is pointless:
unconstrained pin: Warning: There are 18 unconstrained endpoints.
Unconstrained pin: req_rdy
Unconstrained pin: resp_msg[0]
Unconstrained pin: resp_msg[10]
Unconstrained pin: resp_msg[11]
Unconstrained pin: resp_msg[12]

If a user  wants to know about unconstrained endpoints, they can use the opensta check_setup command.

None of the messages printed appear to use the logger. ALL of them should. Like these (because the ones above should not even be there):
Number of end points for restructure 35
Got 1290 pins in cloud extraction
Fount 422 Insts for restructuring ...

jjcherry56 commented:
the submodule commit for opensta in your branch is way out of date (05/30).

when I run it it dies a horrible death that the debugger cannot even tell me where it died, presumably because something barfed on the stack.

Info 1: blif parsed successfully, destroying 422 existing instances...Info 1: inserting 241 new instances...exit
[Thread 0x7fffe6c7e700 (LWP 23340) exited]
[Inferior 1 (process 23328) exited normally]
Missing separate debuginfos, use: debuginfo-install bzip2-libs-1.0.6-13.el7.x86_64 elfutils-libelf-0.176-2.el7.x86_64 elfutils\
-libs-0.176-2.el7.x86_64 freetype-2.8-14.el7.x86_64 glib2-2.56.1-5.el7.x86_64 graphite2-1.3.10-1.el7_3.x86_64 harfbuzz-1.7.5-2\
.el7.x86_64 libX11-1.6.7-2.el7.x86_64 libXau-1.0.8-2.1.el7.x86_64 libXext-1.3.3-3.el7.x86_64 libattr-2.4.46-13.el7.x86_64 libc\
ap-2.22-10.el7.x86_64 libgcrypt-1.5.3-14.el7.x86_64 libglvnd-1.0.1-0.8.git5baa1e5.el7.x86_64 libglvnd-glx-1.0.1-0.8.git5baa1e5\
.el7.x86_64 libgpg-error-1.12-3.el7.x86_64 libicu-50.2-3.el7.x86_64 libpng-1.5.13-7.el7_2.x86_64 libselinux-2.5-14.1.el7.x86_6\
4 libxcb-1.13-1.el7.x86_64 lz4-1.7.5-3.el7.x86_64 ncurses-libs-5.9-14.20130511.el7_4.x86_64 pcre-8.32-17.el7.x86_64 pcre2-utf1\
6-10.23-2.el7.x86_64 python3-libs-3.6.8-18.el7.x86_64 qt5-qtbase-5.9.7-2.el7.x86_64 qt5-qtbase-gui-5.9.7-2.el7.x86_64 readline\
-6.2-11.el7.x86_64 systemd-libs-219-67.el7_7.2.x86_64 tcl-8.5.13-8.el7.x86_64 tcl-tclreadline-2.1.0-3.el7.x86_64 xz-libs-5.2.2\
-1.el7.x86_64 zlib-1.2.7-18.el7.x86_64

jjcherry56 commented:
It looks like the regression write files to /test, which is a no no.
Untracked files (6)
src/rmp/test/gcd_crit_path.blif
src/rmp/test/gcd_crit_path.blif.v
src/rmp/test/gcd_crit_path_out.blif
src/rmp/test/gcd_crit_path_out.blif.v
src/rmp/test/ord_abc_script.tcl

They should be in test/results

mathursanjiv commented:
> the submodule commit for opensta in your branch is way out of date (05/30).
> 
> when I run it it dies a horrible death that the debugger cannot even tell me where it died, presumably because something barfed on the stack.
> 
> Info 1: blif parsed successfully, destroying 422 existing instances...Info 1: inserting 241 new instances...exit
> [Thread 0x7fffe6c7e700 (LWP 23340) exited]
> [Inferior 1 (process 23328) exited normally]
> Missing separate debuginfos, use: debuginfo-install bzip2-libs-1.0.6-13.el7.x86_64 elfutils-libelf-0.176-2.el7.x86_64 elfutils
> -libs-0.176-2.el7.x86_64 freetype-2.8-14.el7.x86_64 glib2-2.56.1-5.el7.x86_64 graphite2-1.3.10-1.el7_3.x86_64 harfbuzz-1.7.5-2
> .el7.x86_64 libX11-1.6.7-2.el7.x86_64 libXau-1.0.8-2.1.el7.x86_64 libXext-1.3.3-3.el7.x86_64 libattr-2.4.46-13.el7.x86_64 libc
> ap-2.22-10.el7.x86_64 libgcrypt-1.5.3-14.el7.x86_64 libglvnd-1.0.1-0.8.git5baa1e5.el7.x86_64 libglvnd-glx-1.0.1-0.8.git5baa1e5
> .el7.x86_64 libgpg-error-1.12-3.el7.x86_64 libicu-50.2-3.el7.x86_64 libpng-1.5.13-7.el7_2.x86_64 libselinux-2.5-14.1.el7.x86_6
> 4 libxcb-1.13-1.el7.x86_64 lz4-1.7.5-3.el7.x86_64 ncurses-libs-5.9-14.20130511.el7_4.x86_64 pcre-8.32-17.el7.x86_64 pcre2-utf1
> 6-10.23-2.el7.x86_64 python3-libs-3.6.8-18.el7.x86_64 qt5-qtbase-5.9.7-2.el7.x86_64 qt5-qtbase-gui-5.9.7-2.el7.x86_64 readline
> -6.2-11.el7.x86_64 systemd-libs-219-67.el7_7.2.x86_64 tcl-8.5.13-8.el7.x86_64 tcl-tclreadline-2.1.0-3.el7.x86_64 xz-libs-5.2.2
> -1.el7.x86_64 zlib-1.2.7-18.el7.x86_64
> 
> So I can't even see an issue with write_sdc because it doesn't get there

In public repo, STA is pointing to "sta @ 370b478". This is what I see in abc_remap branch as well. It looks like there is some mismatch in source code you are seeing and what's there in repo. Some of the other things you see are also old. Hope you are not using abc_remap branch from private repo.
Please check again with below:
git clone --branch abc_remap --recursive https://github.com/mathursanjiv/OpenROAD.git

jjcherry56 commented:
The last think I want to do is clone an entire repo. I added your fork as a remote and then checked out the abc_remap branch but I guess I missed the mark. It worked better the 2nd time. It still doesn't compile because include <string> is missing from blif.h but I fixed that.

Then I get this error:
[ERROR RMP-0020] Cannot open file ./results/ord_abc_script.tcl for writing.

because you are assuming the results dir exists here:
restructure -liberty_file Nangate45/Nangate45_typ.lib -target area -abc_logfile results/abc_rcon.log  -tielo_port $tielo -tiehi_port $tiehi -work_dir ./results

so it will always fail on a cloned repo.
This is why the tcl function make_result_file is defined in helpers.tcl instead of doing things like this:
write_sdc results/out.sdc
that also assume a unix file system syntax instead of using file join

So yes, after all that I do see the seg fault.
Did your remaping intentially delete all but one of the input pins/ports in the def?
If you write the def you can see they are gone.
Which explains why the sdc can't write the name of an input port that it has a pointer to.

mathursanjiv commented:
> The last think I want to do is clone an entire repo. I added your fork as a remote and then checked out the abc_remap branch but I guess I missed the mark. It worked better the 2nd time. It still doesn't compile because include is missing from blif.h but I fixed that.
> 
> Then I get this error:
> [ERROR RMP-0020] Cannot open file ./results/ord_abc_script.tcl for writing.
> 
> because you are assuming the results dir exists here:
> restructure -liberty_file Nangate45/Nangate45_typ.lib -target area -abc_logfile results/abc_rcon.log -tielo_port $tielo -tiehi_port $tiehi -work_dir ./results
> 
> so it will always fail on a cloned repo.
> This is why the tcl function make_result_file is defined in helpers.tcl instead of doing things like this:
> write_sdc results/out.sdc
> that also assume a unix file system syntax instead of using file join
> 
> So yes, after all that I do see the seg fault.
> Did your remaping intentially delete all but one of the input pins/ports in the def?
> If you write the def you can see they are gone.
> Which explains why the sdc can't write the name of an input port that it has a pointer to.

No, we do not intend to delete ports. May be its happening with opendb->blif->ABC->blif->opendb. I'll just check that. Thanks for building, running and helping out.

mathursanjiv commented:
> > The last think I want to do is clone an entire repo. I added your fork as a remote and then checked out the abc_remap branch but I guess I missed the mark. It worked better the 2nd time. It still doesn't compile because include is missing from blif.h but I fixed that.
> > Then I get this error:
> > [ERROR RMP-0020] Cannot open file ./results/ord_abc_script.tcl for writing.
> > because you are assuming the results dir exists here:
> > restructure -liberty_file Nangate45/Nangate45_typ.lib -target area -abc_logfile results/abc_rcon.log -tielo_port $tielo -tiehi_port $tiehi -work_dir ./results
> > so it will always fail on a cloned repo.
> > This is why the tcl function make_result_file is defined in helpers.tcl instead of doing things like this:
> > write_sdc results/out.sdc
> > that also assume a unix file system syntax instead of using file join
> > So yes, after all that I do see the seg fault.
> > Did your remaping intentially delete all but one of the input pins/ports in the def?
> > If you write the def you can see they are gone.
> > Which explains why the sdc can't write the name of an input port that it has a pointer to.
> 
> No, we do not intend to delete ports. May be its happening with opendb->blif->ABC->blif->opendb. I'll just check that. Thanks for building, running and helping out.

Found the issue, deleting net connected to port deletes port. The net deletion was wrong. We'll fix this. Thanks.

maliberty commented:
I've cleaned up a bunch of small issues and the broken build.  The only important issue remaining that I can find is the one liberty versus multiple.  I think we should take that up in a subsequent PR and unblock the rest of this change.  Just waiting on Jenkins now.
